src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -114,17 +114,17 @@ internal sealed partial class ServiceHubRemoteHostClient : RemoteHostClient
             _rpc.StartListening();
         }
 
-        protected override async Task<Session> TryCreateServiceSessionAsync(string serviceName, PinnedRemotableDataScope snapshot, object callbackTarget, CancellationToken cancellationToken)
+        protected override async Task<Session> TryCreateServiceSessionAsync(string serviceName, Func<CancellationToken, Task<PinnedRemotableDataScope>> getSnapshotAsync, object callbackTarget, CancellationToken cancellationToken)
         {
             // get stream from service hub to communicate snapshot/asset related information
             // this is the back channel the system uses to move data between VS and remote host for solution related information
-            var snapshotStream = snapshot == null ? null : await RequestServiceAsync(_hubClient, WellKnownServiceHubServices.SnapshotService, _hostGroup, _timeout, cancellationToken).ConfigureAwait(false);
+            var snapshotStream = getSnapshotAsync == null ? null : await RequestServiceAsync(_hubClient, WellKnownServiceHubServices.SnapshotService, _hostGroup, _timeout, cancellationToken).ConfigureAwait(false);
 
             // get stream from service hub to communicate service specific information
             // this is what consumer actually use to communicate information
             var serviceStream = await RequestServiceAsync(_hubClient, serviceName, _hostGroup, _timeout, cancellationToken).ConfigureAwait(false);
```

Comment:
I would prefer that we do all appropriate work in an area when someone is already in it and focused there.  Otherwise, it's quite likely we'll just "never get around to it" as we tend to often only clean in response to a problem.  Whereas proactive addressing of problems can help us avoid the problems arising in the first place.

Right now it looks like a lot of the code in this area just isn't going to be doing the right thing, and it overlaps the exact area where we're not doing the right hting to the point htat it's causing us an instability problem.  

--

If code is working great, i'm ok with not needing to go and mess with it.  But the issues here seem to have already caused us problems, and that means it's worthwhile to go and try to spend the time to get it in good shape so we don't have to come back to this in the future and page in all the context that we have right now. :)